lib/data/entity.js

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 ODK Central Developers
+// See the NOTICE file at the top-level directory of this distribution and at
+// https://github.com/getodk/central-backend/blob/master/NOTICE.
+// This file is part of ODK Central. It is subject to the license terms in
+// the LICENSE file found in the top-level directory of this distribution and at
+// https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
+// including this file, may be copied, modified, propagated, or distributed
+// except according to the terms contained in the LICENSE file.
+//
+// This is an extension of schema.js where we define *entity-specific* functions
+// for dealing with the XForms XML schema and transforming submission XML into
+// entities.
+
+const csv = require('csv-stringify');
+const { Transform } = require('stream');
+const { PartialPipe } = require('../util/stream');
+const Problem = require('../util/problem');
+const { submissionXmlToFieldStream } = require('./submission');
+
+////////////////////////////////////////////////////////////////////////////
+// ENTITY PARSING
+
+// After collecting the entity data from the submission and form fields,
+// we validate the data (properties must include dataset, label, and uuid)
+// and assemble a valid entity data object. System properties like dataset
+// label, uuid are in entity.system while the contents from the bound form
+// fields are in enttiy.data.
+const validateEntity = (entity) => {
+  const { dataset, label, id } = entity.system;
+
+  if (!dataset || dataset.trim() === '')
+    throw Problem.user.entityViolation({ reason: 'Dataset empty or missing.' });
+
+  if (!label || label.trim() === '')
+    throw Problem.user.entityViolation({ reason: 'Label empty or missing.' });
+
+  if (!id || id.trim() === '')
+    throw Problem.user.entityViolation({ reason: 'ID empty or missing.' });
+
+  const uuidPattern = /^(uuid:)?([0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12})$/i;
+  const matches = uuidPattern.exec(id);
+  if (matches == null) throw Problem.user.entityViolation({ reason: `ID [${id}] not a valid UUID.` });
+  const uuid = matches[2];
```

Comment:
I have been thinking about this too.

> we're going to get the prefix from the client because of how the /meta/entity/id field is set up in the form

We could change what the uid preload does to no longer include the prefix but that would only change what clients do moving forward. I'm starting to think we should consider doing that because the prefix is noisy and I don't think we have any desire to have multiple formats for instanceID and entity id. Still, we'd have to support the prefix for older clients.

Another option would be to use a `setvalue` action on the `odk-instance-first-load` event to set uuids without the prefix. That would only work for clients newer than March 2019. Maybe that's ok? I think it would crash on those old clients so it would be relatively easy to detect (as opposed to silently-missing IDs).